### ISSUE TITLE
Multi-key option with HELib?

### ISSUE THREAD

**Issue by ezamorab** [shaih/HElib#209]
I couldn't find any information about it... Is it possible to use HELib with multi-keys? That is, I want to be able to add values encrypted with different public keys.

**Comment by shaih** [comment]
No, multi-key HE is not implemented HElib (nor in any other library that I know of).

As riku said, you can get a similar effect by using a single public key and secret-sharing the corresponding secret key. This is not very hard if you have a trusted center that can generate the key and then share it.

It is more complicated if you want to generate the key without any one party knowing it. You can look up how to do it here https://eprint.iacr.org/2011/613. But this too is not implemented anywhere that I know of.

**Comment by ezamorab** [comment]
Yes, Riku, I was asking about both cases, either using multiple public keys and a single private key, or using several of each.

Thanks, Shaih. I haven't found a library to implement it with either.

How I could use secret sharing with the private key? Does HELib implement any specific way to do this? Or do I have to simulate it by handling the private key as text and fragment it?  

Thank you in advance.

**Comment by fionser** [comment]
@ezamorab 
* I think HElib now can work with multiple independent sk/pk pairs. What you need to do is just to instantiate multiple FHEcontexts and FHESecKeys.
 * To secrete sharing with the sk, as Shaih has mentioned, you might need a trusted center to split the private key.  One way might to XOR the private key with random sampled strings (something like the one-time padding)

**Comment by shaih** [comment]
HElib does not implement key sharing, but if you have a trusted center that can do it, then you can do the following:

1. Have the center generate a public-key/secret-key as usual

2. If you want to split the key n ways, then choose n-1 random elements modulo q, call them s1.s2,...,s_{n-1}. Then compute s_n = secKey.sKeys[0] - \sum_i s_i (mod q). Give si to the i'th party.

3. When you have a ciphertext (c0,c1) to decrypt (relative to keyID=0 , modulus q, and plaintext space mod p), each  party i computes a share sh_i = c1*si + ei*p (mod q) (for some small random error ei) and broadcasts sh_i to everyone. The decrypted message is computed as

  m = (c0 + \sum_i sh_i mod q) mod p

Security of this procedure depends on c1 being random (at least "random enough"), if you let an adversary give you arbitrary c1's then it can recover the secret-key shares si's.

**Comment by ezamorab** [comment]
Thank you for your answers.

@shaih I have read your answer several times, Shai, but I cannot fully understand it... I have tried to get on with it and start developing it, but I really don't know how I should proceed... 
Are there any examples I can take as a guide? I find it really useful to use secret sharing and would like to implement it in my project, but perhaps it is also too difficult.

@fionser You said that HELib can work with multiple independent sk/pk pairs.
But with it, could I compute (sum) encrypted data from different contexts and decipher the result?

**Comment by shaih** [comment]
Consider the code in `FHESecKey::ImportSecKey(const DoubleCRT& sKey,...)` (in `FHE.cpp`): You need to do something like this:
```
DoubleCRT accumulator(context); // initalized to zero
for (long i=0; i<n-1; i++) {
  DoubleCRT si(context);
  si.randomize(); // get a random element modulo q
  accumulator -= si;
  // ... send si to the i'th party
}
accumulator += sKey; // compute the last key
// .. send si=accumulator to the last party
```
For all of these "decrypting parties", you will have to keep the received si, maybe as a `DoubleCRT`. To decrypt, each decrypting party can call for example the method `RLWE(share_i, ctxt.parts[1], si, ctxt.getPtxtSpace())` (also in `FHE.cpp`). The result is the decryption share (or maybe the negation of it). Then you again do an accumulation of all the shares, add to it also `ctxt.parts[0]`, then reduce modulo p.
